### PR TITLE
Schema manager should handle shard key on reference

### PR DESCRIPTION
**Pull request by notrix**
Schema manager should also know how to handle shard key on reference
This PR is part of sharding key on reference fix. Referencing another PR #1687

### CODE REVIEW

**Comment by darceee2** [comment]
:+1:

File: lib/Doctrine/ODM/MongoDB/SchemaManager.php [hunk 1]
```diff
@@ -609,10 +610,32 @@ private function runShardCollectionCommand($documentName)
         $shardKey = $class->getShardKey();
         $adminDb = $this->dm->getConnection()->selectDatabase('admin');
 
+        $shardKeyPart = array();
+        foreach ($shardKey['keys'] as $key => $order) {
+            try {
+                $mapping = $class->getFieldMappingByDbFieldName($key);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/SchemaManager.php]: I'll have to investigate field name translation here - there may be a couple of things wrong with field name translations in this method, unrelated to your patch.

File: lib/Doctrine/ODM/MongoDB/SchemaManager.php [hunk 2]
```diff
@@ -609,10 +610,32 @@ private function runShardCollectionCommand($documentName)
...
+                if (isset($mapping['association']) && $mapping['association'] === ClassMetadata::REFERENCE_ONE) {
+                    switch ($mapping['storeAs']) {
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/SchemaManager.php]: Use `ClassMetadataInfo::getReferenceId` instead of duplicating logic.

Comment (notrix) [reply lib/Doctrine/ODM/MongoDB/SchemaManager.php]: I could not find how to use ClassMetadataInfo::getReferenceId() as I don't have reference mongo data. But I have found ClassMetadataInfo::getReferenceFieldName() useful in this case.

Comment (alcaeus) [reply lib/Doctrine/ODM/MongoDB/SchemaManager.php]: Yes, my apologies - got confused. `getReferenceFieldName` is correct.

**Comment by alcaeus** [comment]
@notrix after an eternity without updates, I finally managed to revisit this. I've created #1748 to fix this issue in the 1.2 branch, to be released in 1.2.2 soon.

I've rebased this branch on top of the current master to spare you the trouble - with the `ext-mongodb` and CS merges lots of things have changed. Due to the `ext-mongodb` change, sharding support in master is currently incomplete, so I've skipped the test for now. I'll revisit this in #1698 to work around the issues preventing finalizing this. Once #1748 has been merged, I'll also merge this branch down.

Thanks for your contribution! 🎉

File: tests/Documents/Sharded/ShardedByUser.php [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+declare(strict_types=1);
+
+namespace Documents\Sharded;
+
+use Doctrine\ODM\MongoDB\Mapping\Annotations as ODM;
+
+/**
+ * @ODM\Document(collection="sharded.users")
+ * @ODM\ShardKey(keys={"user"="asc"})
```
Comment (alcaeus) [inline tests/Documents/Sharded/ShardedByUser.php]: FYI @notrix, I had to change this from `desc` to `asc` as the test failed with the error that the shard key would could only be ascending or hashed.

**Pull request merged**